File: frontend/src/api/notify.ts
import request from './request'

/**
 * 通知中心 API（M09；dev-plan P9.1/P9.2/P9.3；api-spec §4.9）。
 */

export type NotifyLevel = 'info' | 'warn' | 'error'
export type ChannelType = 'bark' | 'telegram' | 'smtp' | 'webhook' | 'wecom' | 'dingtalk' | 'feishu' | 'ntfy'
export type NotifyEvent =
  | 'app_down'
  | 'app_up'
  | 'metric_alert'
  | 'port_down'
  | 'port_up'
  | 'flow_failed'
  | 'system'

export interface NotificationItem {
  id: number
  title: string
  body: string
  level: NotifyLevel
  source: string
  is_read: boolean
  created_at: string
}

export interface NotificationPage {
  items: NotificationItem[]
  total: number
  unread: number
}

export interface NotifyChannel {
  id: number
  type: ChannelType
  name: string
  enabled: boolean
  config: Record<string, unknown>
}

export interface NotifyRule {
  id: number
  event: NotifyEvent
  channel_ids: number[]
  enabled: boolean
  quiet_start: string | null
  quiet_end: string | null
}

export const notifyApi = {
  list: (params: { level?: string; unread?: number; limit?: number; offset?: number } = {}) =>
    request.get<never, NotificationPage>('/notifications', { params }),
  unreadCount: () => request.get<never, { unread: number }>('/notifications/unread-count'),
  markRead: (id: number) => request.put(`/notifications/${id}/read`),
  readAll: () => request.put('/notifications/read-all'),
  remove: (id: number) => request.delete(`/notifications/${id}`),

  channels: () => request.get<never, NotifyChannel[]>('/notify-channels'),
  createChannel: (body: Omit<NotifyChannel, 'id'>) => request.post<never, NotifyChannel>('/notify-channels', body),
  updateChannel: (id: number, body: Omit<NotifyChannel, 'id'>) =>
    request.put<never, NotifyChannel>(`/notify-channels/${id}`, body),
  deleteChannel: (id: number) => request.delete(`/notify-channels/${id}`),
  testChannel: (id: number) => request.post<never, { sent: boolean }>(`/notify-channels/${id}/test`),

  rules: () => request.get<never, NotifyRule[]>('/notify-rules'),
  replaceRules: (rules: Array<Omit<NotifyRule, 'id'>>) => request.put<never, NotifyRule[]>('/notify-rules', { rules }),
}

File: backend/app/models/notify.py
"""通知渠道与路由规则（M09-4~12；dev-plan P9.1/P9.3；api-spec §3.8）。

- notify_channels：type 枚举 bark/telegram/smtp/webhook/wecom/dingtalk/feishu/ntfy；
  config 为 JSON，敏感字段回传时脱敏（见 api/v1/notify.py）；
- notify_rules：事件 → 渠道路由；quiet_start/quiet_end 规则级免打扰（HH:MM，可跨午夜）。
"""

from __future__ import annotations

from datetime import datetime

from sqlalchemy import DateTime, Integer, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base

CHANNEL_TYPES = ("bark", "telegram", "smtp", "webhook", "wecom", "dingtalk", "feishu", "ntfy")
NOTIFY_EVENTS = (
    "app_down",
    "app_up",
    "metric_alert",
    "port_down",
    "port_up",
    "flow_failed",
    "system",
)


class NotifyChannel(Base):
    __tablename__ = "notify_channels"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    type: Mapped[str] = mapped_column(Text, default="webhook")
    name: Mapped[str] = mapped_column(Text, default="")
    config: Mapped[str] = mapped_column(Text, default="{}")  # JSON；敏感字段回传脱敏
    enabled: Mapped[int] = mapped_column(Integer, default=1)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)


class NotifyRule(Base):
    __tablename__ = "notify_rules"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    event: Mapped[str] = mapped_column(Text, default="system")
    channel_ids: Mapped[str] = mapped_column(Text, default="[]")  # JSON 数组
    enabled: Mapped[int] = mapped_column(Integer, default=1)
    quiet_start: Mapped[str | None] = mapped_column(Text, default=None)  # HH:MM
    quiet_end: Mapped[str | None] = mapped_column(Text, default=None)  # HH:MM
